### PR TITLE
Lsq solver

### PR DESCRIPTION
**Pull request by Hebbalali**
Add the Least Squares Solver (LsqSolver) for the resolution of a system of non-linear equations.
The implementation uses the the least squares optimization algorithm from Dlib.

### CODE REVIEW

File: lib/src/Base/Solver/LsqSolver.cxx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+Point LsqSolver::solve(const Function & function,
+                     const Point & startingPoint) const
+{ 
+  UnsignedInteger callsNumber = 0;
+  const UnsignedInteger maximumCallsNumber = getMaximumCallsNumber();
+  const Scalar absoluteError = getAbsoluteError();
+  const Scalar relativeError = getRelativeError();
+  const Scalar residualError = getResidualError();
+  LeastSquaresProblem lsqProblem(function);
+  Dlib lsqAlgorithm(lsqProblem, "least_squares");
```
Comment (jschueller) [inline lib/src/Base/Solver/LsqSolver.cxx:79]: we dont want to hardcode the Dlib solver here depending on the configuration
instead, OptimizationAlgorithm::Build(OptimizationProblem) method can select automatically a least-squares solver available; just pass the lsqProblem and you get a suited OptimizationAlgorithm in return

File: lib/src/Base/Solver/openturns/SolverImplementation.hxx [hunk 2]
```diff
@@ -90,6 +90,10 @@ public:
                        const Scalar infValue,
                        const Scalar supValue) const;
 
+  /** Solve attempt to find one root to a system of equations function(x) = 0 given a starting point x_0 */
+  virtual Point solve(const Function & function,
+                      const Point & startingPoint) const;
```
Comment (jschueller) [inline lib/src/Base/Solver/openturns/SolverImplementation.hxx:95]: maybe we could add an extra OT::Interval argument (eventually empty by default) as the 1-d methods do

File: lib/src/Base/Solver/openturns/LsqSolver.hxx [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+
+BEGIN_NAMESPACE_OPENTURNS
+
+/**
+ * @class LsqSolver
+ *
+ * This class is an interface for the 1D nonlinear solverImplementations
+ */
+class OT_API LsqSolver :
```
Comment (jschueller) [inline lib/src/Base/Solver/openturns/LsqSolver.hxx:35]: LsqSolver is probably ambiguous as we have already lots of LeastSquares* classes
maybe LeastSquaresEquationSolver ? LeastSquaresRootSolver ?
maybe @regislebrun has a better name for this

File: python/src/solver_module.i [hunk 4]
```diff
@@ -53,6 +53,7 @@
 %include ODESolver.i
 %include RungeKutta.i
 %include Fehlberg.i
+%include LsqSolver.i
```
Comment (jschueller) [inline python/src/solver_module.i:56]: new classes must go to experimental_module.i instead (correct the CMakeLists accordingly)

File: python/src/LsqSolver_doc.i.in [hunk 5]
```diff
@@ -0,0 +1,55 @@
+%feature("docstring") OT::LsqSolver
+"Least Square Solver (LsqSolver) for a system of non-linear equations.
+
+Available constructor:
+    LsqSolver()
```
Comment (jschueller) [inline python/src/LsqSolver_doc.i.in:5]: no need to document the default ctor, so you can drop this "Available constructor"

File: lib/src/Base/Solver/SolverImplementation.cxx [hunk 6]
```diff
@@ -236,4 +236,11 @@ Scalar SolverImplementation::solve(const UniVariateFunction &,
   throw NotYetImplementedException(HERE) << "In SolverImplementation::solve(const UniVariateFunction & function, const Scalar value, const Scalar infPoint, const Scalar supPoint, const Scalar infValue, const Scalar supValue) const";
 }
 
+/** Solve attempt to find one root to a system of equations function(x) = 0 given a starting point x_0 */
+Point SolverImplementation::solve(const Function &,
+                                  const Point &) const
+{
+  throw NotYetImplementedException(HERE) << "In SolverImplementation::solve(const UniVariateFunction & function, const Scalar value, const Scalar infPoint, const Scalar supPoint, const Scalar infValue, const Scalar supValue) const";
```
Comment (jschueller) [inline lib/src/Base/Solver/SolverImplementation.cxx:243]: the message does not match the arguments

**Comment by jschueller** [comment]
nice first contribution!

File: lib/src/Base/Solver/LsqSolver.cxx (hunk 1 above)
Comment (Hebbalali) [reply lib/src/Base/Solver/LsqSolver.cxx]: By default now it uses the least squares implementation in Ceres (by default it goes to the first alphabetically compatible solver)

File: lib/src/Base/Solver/openturns/LsqSolver.hxx (hunk 3 above)
Comment (Hebbalali) [reply lib/src/Base/Solver/openturns/LsqSolver.hxx:35]: I think we can go for LeastSquaresEquationsSolver, meaning that it's suitable also for a system of equations.
Question: to rename a class in the library, we use the same "customize.sh" that is used for module renaming?

File: lib/src/Base/Solver/openturns/SolverImplementation.hxx (hunk 2 above)
Comment (Hebbalali) [reply lib/src/Base/Solver/openturns/SolverImplementation.hxx:95]: I don't think that I fully understand what you mean here. Do you mean to add an infPoint and supPoint arguments in this method as well?

Comment (jschueller) [reply lib/src/Base/Solver/openturns/SolverImplementation.hxx:95]: an OT::Interval would be equivalent to adding infPoint/supPoint but would allow for infinite intervals too, also it can be passed directly to the OptimizationAlgorithm

File: python/src/LsqSolver_doc.i.in [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+LsqSolver might fail and not obtain a result lower than the specified threshold, in this case an error is thrown."
+
+// ---------------------------------------------------------------------
+
+%feature("docstring") OT::LsqSolver::solve
+"Solve a system of non-linear equations.
+
+**Available usages**:
+
+    solve(*function, startingPoint*)
```
Comment (jschueller) [inline python/src/LsqSolver_doc.i.in:34]: you can drop this "Available usages" section

File: python/src/LsqSolver_doc.i.in [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+Parameters
+----------
+function : :class:`~openturns.Function`
+    The vectorial function of the system of equations :math:`function_i(x) = 0 \forall i \in {1,...,n}` to be solved. Remark that it is an unbounded resolution.
+startingPoint : Point
+    The starting point of the optimization algorithm.
+
+Returns
+-------
+result : Point
```
Comment (jschueller) [inline python/src/LsqSolver_doc.i.in:45]: :class:`~openturns.Point`

File: python/src/LsqSolver_doc.i.in [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+**Available usages**:
+
+    solve(*function, startingPoint*)
+
+Parameters
+----------
+function : :class:`~openturns.Function`
+    The vectorial function of the system of equations :math:`function_i(x) = 0 \forall i \in {1,...,n}` to be solved. Remark that it is an unbounded resolution.
+startingPoint : Point
```
Comment (jschueller) [inline python/src/LsqSolver_doc.i.in:40]: sequence of float

File: lib/src/Base/Solver/openturns/LsqSolver.hxx (hunk 3 above)
Comment (sofianehaddad) [reply lib/src/Base/Solver/openturns/LsqSolver.hxx:35]: There is a `rename.sh` in `utils`

Usage :  `./utils/rename.sh OldClass NewClass `

File: .github/workflows/build.yml [hunk 10]
```diff
@@ -11,7 +11,6 @@ jobs:
     steps:
     - uses: actions/checkout@v4
     - run: |
-        brew upgrade
```
Comment (mbaudin47) [inline .github/workflows/build.yml:14]: I do not think that you should update this file in this PR. Is that really necessary?

**Review by mbaudin47 — COMMENTED** [review]
Very nice feature! It requires a few changes though.

File: lib/src/Base/Solver/openturns/LeastSquaresEquationsSolver.hxx [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+  /** Virtual constructor */
+  LeastSquaresEquationsSolver * clone() const override;
+
+  /** String converter */
+  String __repr__() const override;
+
+  /** Solve attempt to find one root to a system of equations function(x) = 0 given a starting point x_0 */
+  using SolverImplementation::solve;
+  Point solve(const Function & function,
+               const Point & startingPoint) const override;
```
Comment (mbaudin47) [inline lib/src/Base/Solver/openturns/LeastSquaresEquationsSolver.hxx:57]:
```suggestion
              const Point & startingPoint) const override;
```

File: lib/src/Base/Solver/openturns/Solver.hxx [hunk 12]
```diff
@@ -78,6 +78,10 @@ public:
                        const Scalar infValue,
                        const Scalar supValue) const;
 
+  /** Solve attempt to find one root to a system of equations function(x) = 0 given a starting point x_0 */
+  virtual Point solve(const Function & function,
+                       const Point & startingPoint) const;
```
Comment (mbaudin47) [inline lib/src/Base/Solver/openturns/Solver.hxx:83]:
```suggestion
                      const Point & startingPoint) const;
```

File: lib/test/t_LeastSquaresEquationsSolver_std.cxx [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+int main(int, char *[])
+{
+  TESTPREAMBLE;
+  OStream fullprint(std::cout);
+
+  try
+  {
+    /** Analytical construction */
+    Description input = {"x","y"};
+    Description formulas = {"y*x-sin(2*x)","1 + cos(y) + x"};
```
Comment (mbaudin47) [inline lib/test/t_LeastSquaresEquationsSolver_std.cxx:36]:
```suggestion
    Description formulas = {"y * x - sin(2 * x)","1 + cos(y) + x"};
```

File: lib/test/t_LeastSquaresEquationsSolver_std.cxx [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+
+    SymbolicFunction analytical(input, formulas);
+    LeastSquaresEquationsSolver myAlgo;
+    myAlgo.setResidualError(1e-5);
+    fullprint << "myAlgo=" << myAlgo << std::endl;
+    Point startingPoint = {2.0,1.0};
```
Comment (mbaudin47) [inline lib/test/t_LeastSquaresEquationsSolver_std.cxx:42]:
```suggestion
    Point startingPoint = {2.0, 1.0};
```

File: python/src/DistributionImplementation_doc.i.in [hunk 15]
```diff
@@ -15,13 +15,13 @@ OT_Distribution_doc
 
 Parameters
 ----------
-X : sequence of float, 2-d sequence of float
-    CDF input(s).
+x : sequence of float, 2-d sequence of float
```
Comment (mbaudin47) [inline python/src/DistributionImplementation_doc.i.in:18]: These updates are interesting, but they are not associated to the non linear equation solver that this PR is about. I suggest to create a separate PR to discuss them.

File: lib/test/t_LeastSquaresEquationsSolver_std.cxx [hunk 16]
```diff
@@ -0,0 +1,56 @@
...
+    OSS oss;
+    oss.setPrecision(5);
+    fullprint << "x=" << String(oss << myAlgo.solve(analytical, startingPoint)) << std::endl;
```
Comment (mbaudin47) [inline lib/test/t_LeastSquaresEquationsSolver_std.cxx:45]: Unit tests based on expected outputs tend to be difficult to maintain. Please use assert features for ottest instead. Does the exact solution of this problem is known? If not, I suggest to use [Rosenbrock's test problem](https://en.wikipedia.org/wiki/Rosenbrock_function) which has exactly the same mathematical structure, i.e. a system of two equations solved by least squares (plus, the exact solution is known).

File: lib/test/t_LeastSquaresEquationsSolver_std.expout [hunk 17]
```diff
@@ -0,0 +1,40 @@
+Distribution class=MarginalDistribution name=MarginalDistribution dimension=3 distribution=class=Normal name=Normal dimension=5 mean=class=Point name=Unnamed dimension=5 values=[0,0,0,0,0] sigma=class=Point name=Unnamed dimension=5 values=[1,1,1,1,1] correlationMatrix=class=CorrelationMatrix dimension=5 implementation=class=MatrixImplementation name=Unnamed rows=5 columns=5 values=[1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,1] indices=[2,0,1]
```
Comment (mbaudin47) [inline lib/test/t_LeastSquaresEquationsSolver_std.expout:1]: Provided you use ottest, this file won't be necessary anymore.

File: lib/test/t_LeastSquaresEquationsSolver_std.cxx [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+    fullprint << "myAlgo=" << myAlgo << std::endl;
+  }
+  catch (TestFailed & ex)
+  {
+    std::cerr << ex << std::endl;
+    return ExitCode::Error;
+  }
+
```
Comment (mbaudin47) [inline lib/test/t_LeastSquaresEquationsSolver_std.cxx:53]:
```suggestion
```

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 19]
```diff
@@ -0,0 +1,47 @@
+%feature("docstring") OT::LeastSquaresEquationsSolver
+"Least Square Solver (LeastSquaresEquationsSolver) for a system of non-linear equations.
+
+
+Parameters
+----------
+absError : positive float
+    Absolute error: distance between two successive iterates at the end point.
+    Default is :math:`10^{-5}`.
```
Comment (mbaudin47) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:9]: That is not true. The default value is from the `ResourceMap` key "Solver-DefaultAbsoluteError". Users should know this.

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 20]
```diff
@@ -0,0 +1,47 @@
+%feature("docstring") OT::LeastSquaresEquationsSolver
+"Least Square Solver (LeastSquaresEquationsSolver) for a system of non-linear equations.
+
+
+Parameters
+----------
+absError : positive float
+    Absolute error: distance between two successive iterates at the end point.
```
Comment (mbaudin47) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:8]: The significant fact is that this is an absolute tolerance on x, not f(x) isn'it?
```suggestion
    Absolute error: distance between two successive x iterates at the end point.
```

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 21]
```diff
@@ -0,0 +1,47 @@
...
+    Default is :math:`10^{-5}`.
+relError : positive float
+    Relative error: distance between the two last successive iterates with
+    regards to the last iterate. Default is :math:`10^{-5}`.
```
Comment (mbaudin47) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:12]:
```suggestion
    Relative error: relative distance between the two last successive iterates x with
    respect to the last iterate. Default is :math:`10^{-5}`.
```
Please introduce the associated ResourceMap key here too.

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 22]
```diff
@@ -0,0 +1,47 @@
...
+resError : positive float
+    Residual error: threshold for accepting a given solution. It is also used as residual error for the least square optimizer.
+    value. Default is :math:`0`.
```
Comment (mbaudin47) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:15]: Please introduce the associated ResourceMap key here too.

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 23]
```diff
@@ -0,0 +1,47 @@
...
+Parameters
+----------
+absError : positive float
+    Absolute error: distance between two successive iterates at the end point.
+    Default is :math:`10^{-5}`.
+relError : positive float
+    Relative error: distance between the two last successive iterates with
+    regards to the last iterate. Default is :math:`10^{-5}`.
+resError : positive float
+    Residual error: threshold for accepting a given solution. It is also used as residual error for the least square optimizer.
```
Comment (mbaudin47) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:14]:
```suggestion
    Residual error: absolute threshold on f(x) for accepting a given solution. It is also used as residual error for the least square optimizer.
```

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+    value. Default is :math:`0`.
+maximumFunctionEvaluation : int
+    The maximum number of evaluations of the function. Default is :math:`100`.
```
Comment (mbaudin47) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:17]: Please introduce the associated ResourceMap key here too.

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+
+See also
+--------
+openturns.Solver, openturns.OptimizationAlgorithm
+
+Notes
+-----
+LeastSquaresEquationsSolver might fail and not obtain a result lower than the specified threshold, in this case an error is thrown."
```
Comment (mbaudin47) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:25]: Please limit yourself to 79 characters within a line, because this can be difficult to review and maintain otherwise.
```suggestion
LeastSquaresEquationsSolver might fail and not obtain a result lower than the
specified threshold, in this case an error is thrown."
```

File: python/src/LeastSquaresEquationsSolver_doc.i.in [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+    The starting point of the optimization algorithm.
+
+Returns
+-------
+result : :class:`~openturns.Point`
+    The result of the root research.
+
+Notes
+-----
+LeastSquaresEquationsSolver might fail and not obtain a result lower than the specified threshold, in this case an error is thrown."
```
Comment (mbaudin47) [inline python/src/LeastSquaresEquationsSolver_doc.i.in:47]:
```suggestion
LeastSquaresEquationsSolver might fail and not obtain a result lower than the
specified threshold, in this case an error is thrown."
```

File: python/test/t_LeastSquaresEquationsSolver_std.py [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+myAlgo = otexp.LeastSquaresEquationsSolver()
+myAlgo.setResidualError(1e-10)
+myAlgo.setMaximumCallsNumber(1000)
+print("myAlgo=", myAlgo)
+starting_point = ot.Point([2.0, 1.0])
+print("Solve ", formulas, "= [0,0] for ", input)
+solution = myAlgo.solve(analytical, starting_point)
+print("[x,y] = ", solution)
+print("myAlgo=", myAlgo)
+ott.assert_almost_equal(analytical(solution), [0, 0], 1e-5, 1e-5)
```
Comment (mbaudin47) [inline python/test/t_LeastSquaresEquationsSolver_std.py:24]: That is a perfect unit test. Please use the same assert pattern in the C++.

**Review by mbaudin47 — CHANGES_REQUESTED** [review]
Thank you for sharing such an interesting PR! This feature is highly missing in the library, indeed. Don't you feel that this could be named `NonLinearEquationSolver`? This is because this is the feature provided by the code, independently of its actual implementation. The implementation may be updated afterwards, based on dill, as you suggested.

File: python/src/solver_module.i [hunk 28]
```diff
@@ -53,6 +53,5 @@
 %include ODESolver.i
 %include RungeKutta.i
 %include Fehlberg.i
-
```
Comment (mbaudin47) [inline python/src/solver_module.i:56]: That is an unnecessary change.

**Comment by mbaudin47** [comment]
The Linter reveals to bugs:
python/doc/user_manual/threshold_probability_simulation_algorithms.rst:79: D002 Trailing whitespace
python/doc/user_manual/threshold_probability_simulation_algorithms.rst:84: D002 Trailing whitespace

Please rebase on top of master, because lots of code have been merged recently.

File: python/doc/user_manual/threshold_probability_simulation_algorithms.rst [hunk 29]
```diff
@@ -75,11 +75,13 @@ Non linear solvers
     :template: class.rst_t
 
     Solver
-
+    experimental.LeastSquaresEquationsSolver
+
```
Comment (mbaudin47) [inline python/doc/user_manual/threshold_probability_simulation_algorithms.rst:79]: This is a point where Lint points an extra whitespace.

File: python/doc/user_manual/threshold_probability_simulation_algorithms.rst [hunk 30]
```diff
@@ -75,11 +75,13 @@ Non linear solvers
...
 
     Solver
-
+    experimental.LeastSquaresEquationsSolver
+    
     :template: Solver.rst_t
     Bisection
     Brent
     Secant
+
```
Comment (mbaudin47) [inline python/doc/user_manual/threshold_probability_simulation_algorithms.rst:84]: This is a point where Lint points an extra whitespace.

File: python/src/LeastSquaresEquationsSolver_doc.i.in (hunk 24 above)
Comment (Hebbalali) [reply python/src/LeastSquaresEquationsSolver_doc.i.in:17]: Already existing. It is a bad copy/paste from `Brent` doc. This will be improved for this class

File: lib/src/Base/Solver/openturns/SolverImplementation.hxx (hunk 2 above)
Comment (sofianehaddad) [reply lib/src/Base/Solver/openturns/SolverImplementation.hxx:95]: @jschueller done. For the moment two different methods but probably will use the hint of an empty interval

File: python/src/LeastSquaresEquationsSolver_doc.i.in (hunk 22 above)
Comment (sofianehaddad) [reply python/src/LeastSquaresEquationsSolver_doc.i.in:15]: Done. The initial doc is a copy/paste from `Secant`, `Bidirectional` which have to be reviewed later
